docs/docsite/rst/dev_guide/developing_plugins.rst

```diff
@@ -477,62 +477,97 @@ Vars plugins
+                   cache_key = entity.name + entity.base_type
+               except AttributeError:
+                   raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
 
- * loader: Ansible's DataLoader. The DataLoader can read files, auto-load JSON/YAML and decrypt vaulted data, and cache read files.
- * path: this is 'directory data' for every inventory source and the current play's playbook directory, so they can search for data in reference to them. ``get_vars`` will be called at least once per available path.
- * entities: these are host or group names that are pertinent to the variables needed. The plugin will get called once for hosts and again for groups.
+               if cache_key not in FOUND:
+                   FOUND[cache_key] = find_variable_source_for_entity(path, entity)
+
+               variables.update(self.load_variables_from_source(loader, FOUND[cache_key]))
+           return variables
 
-This ``get_vars`` method just needs to return a dictionary structure with the variables.
+In addition to running many times, a vars plugin is instantiated by default every time it might run. Stateless plugins like ``host_group_vars`` can opt into reusing the same instance throughout a play by setting the ``cache_instance`` class attribute to True.
+
+.. code-block:: python
 
-Since Ansible version 2.4, vars plugins only execute as needed when preparing to execute a task. This avoids the costly 'always execute' behavior that occurred during inventory construction in older versions of Ansible. Since Ansible version 2.10, vars plugin execution can be toggled by the user to run when preparing to execute a task or after importing an inventory source.
+   class VarsModule(BaseVarsPlugin):
 
-The user must explicitly enable vars plugins that reside in a collection. See :ref:`enable_vars` for details.
+       cache_instance = True
 
-Legacy vars plugins are always loaded and run by default. You can prevent them from automatically running by setting ``REQUIRES_ENABLED`` to True.
+Vars plugins in non-collection search paths run by default without user-configuration. To prevent this, set the ``REQUIRES_ENABLED`` class attribute to True, and document that the legacy plugin needs to be enabled to run.
 
 .. code-block:: python
 
-    class VarsModule(BaseVarsPlugin):
-        REQUIRES_ENABLED = True
+   DOCUMENTATION = '''
+        name: legacy_vars_plugin
+        notes:
+          - This plugin must be enabled to run.
+   '''
 
-Include the ``vars_plugin_staging`` documentation fragment to allow users to determine when vars plugins run.
+   class VarsModule(BaseVarsPlugin):
+
+       REQUIRES_ENABLED = True
+
+Historically, vars plugins have had a few changes in implementation and behavior.
+
+Vars plugins were partially implemented in Ansible 2.0 and rewritten to be fully implemented starting with Ansible 2.4. Vars plugins are supported by collections starting with Ansible 2.10.
+
```

Comment:
Given we are on 2.15, do we need to keep the historical info here? Is it of value to anyone creating plugins today?

Reply:
I could delete one sentence:
```
Vars plugins were partially implemented in Ansible 2.0 and rewritten to be fully implemented starting with Ansible 2.4. 
```

But the historical example below still works for example and hasn't been deprecated. If ansible collections still support 2.9, I think it's worth keeping the note about when vars plugins are supported in collections too.

Reply:
good point, should we take this as the opportunity to deprecate 1.0 vars plugins?

Reply:
yeah 2.9 is still relevant for sure. thanks!

Reply:
I was mistaken about the example below still working - we actually removed v1 vars plugins a while ago so I just removed that example. We never added developer documentation for (v2?) vars plugins using `get_host_vars`/`get_group_vars` instead of `get_vars`, so I added a deprecation for that.